src/invariant/InvariantTests.cpp

```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+    st.exchange(soci::into(lastModified));
+    st.define_and_bind();
+    st.execute(true);
+    while (st.got_data())
+    {
+        allLastModified.push_back(lastModified);
+        st.fetch();
+    }
+    return allLastModified;
+}
+
+class ApplyBucketsWorkAddEntry : public ApplyBucketsWork
+{
+  private:
+    uint32_t mFromLedgerSeq;
+    LedgerEntryType mType;
+    bool mAdded;
+
+  public:
+    ApplyBucketsWorkAddEntry(
+            Application& app, WorkParent& parent,
+            std::map<std::string, std::shared_ptr<Bucket>> const& buckets,
+            HistoryArchiveState const& applyState,
+            uint32_t addAtLedgerSeq,
+            LedgerEntryType addType)
+        : ApplyBucketsWork(app, parent, buckets, applyState)
+        , mFromLedgerSeq{addAtLedgerSeq}
+        , mType{addType}
+        , mAdded{false}
+    {
+    }
+
+    Work::State onSuccess() override
+    {
+        if (!mAdded)
+        {
+            auto& db = mApp.getDatabase();
+            size_t count = 0;
+            {
+                auto prep = db.getPreparedStatement(
+                    "SELECT"
+                    " (SELECT COUNT(*) AS c FROM accounts WHERE lastmodified= :v1)"
+                    " + (SELECT COUNT(*) AS c FROM trustlines WHERE lastmodified= :v1)"
+                    " + (SELECT COUNT(*) AS c FROM offers WHERE lastmodified= :v1)"
+                    " + (SELECT COUNT(*) AS c FROM accountdata WHERE lastmodified= :v1)");
+                auto& st = prep.statement();
+                st.exchange(soci::into(count));
+                st.exchange(soci::use(mFromLedgerSeq));
+                st.define_and_bind();
+                st.execute(true);
+            }
+
+            if (count > 0)
+            {
+                LedgerEntry entry;
+                entry.data.type(mType);
+                entry.lastModifiedLedgerSeq = mFromLedgerSeq;
+                switch (mType)
+                {
+                case ACCOUNT:
+                    entry.data.account() = LedgerTestUtils::generateValidAccountEntry(5);
+                    break;
+                case TRUSTLINE:
+                    entry.data.trustLine() = LedgerTestUtils::generateValidTrustLineEntry(5);
+                    break;
+                case OFFER:
+                    entry.data.offer() = LedgerTestUtils::generateValidOfferEntry(5);
+                    break;
+                case DATA:
+                    entry.data.data() = LedgerTestUtils::generateValidDataEntry(5);
+                    break;
+                default:
+                    REQUIRE(false);
+                }
+
+                LedgerHeader lh;
+                LedgerDelta ld(lh, db, false);
+                auto ef = EntryFrame::FromXDR(entry);
+                ef->storeAdd(ld, db);
+                mAdded = true;
+            }
+        }
+        return ApplyBucketsWork::onSuccess();
+    }
+};
+
+class ApplyBucketsWorkDeleteEntry : public ApplyBucketsWork
```

Comment:
I think you could also consider making one class that accepts lambda executed once on onSuccess here instead of 3 classes.